### PR TITLE
[SPARK-48463][ML] Make StringIndexer supporting nested input columns

### PR DESCRIPTION
**Pull request by WeichenXu123**
### What changes were proposed in this pull request?

Make StringIndexer supporting nested input columns

### Why are the changes needed?

User demand.

### Does this PR introduce _any_ user-facing change?

Yes.

### How was this patch tested?

Unit tests.

### CODE REVIEW

File: mllib/src/main/scala/org/apache/spark/ml/feature/StringIndexer.scala [hunk 1]
```diff
@@ -124,14 +124,34 @@ private[feature] trait StringIndexerBase extends Params with HasHandleInvalid wi
...
+          )
+          case None if skipNonExistsCol => None
           case _ => throw new SparkException(s"Input column $inputColName does not exist.")
         }
     }
     StructType(schema.fields ++ outputFields)
   }
+
+  protected def extractInputDataType(schema: StructType, inputColName: String): Option[DataType] = {
+    val inputSplits = inputColName.split("\\.")
```
Comment (HyukjinKwon) [inline mllib/src/main/scala/org/apache/spark/ml/feature/StringIndexer.scala:139]: I think you can use `UnresolvedAttribute.parseAttributeName`

Comment (zhengruifeng) [reply mllib/src/main/scala/org/apache/spark/ml/feature/StringIndexer.scala:139]: +1

**Comment by zhengruifeng** [comment]
is it possible to extract this field first by `df.select('a.b.c')` ?

**Comment by WeichenXu123** [comment]
> is it possible to extract this field first by `df.select('a.b.c')` ?

I don't like this way, in this way we have to catch exception raised by `df.select('a.b.c')` and match error message , it leads to fragile code.

**Comment by zhengruifeng** [comment]
> > is it possible to extract this field first by `df.select('a.b.c')` ?
> 
> I don't like this way, in this way we have to catch exception raised by `df.select('a.b.c')` and match error message , it leads to fragile code.

I still hesitate about this approach, since there maybe some related configurations, e.g. case sensitive or not

**Review by zhengruifeng — APPROVED** [review]

File: mllib/src/main/scala/org/apache/spark/ml/feature/StringIndexer.scala [hunk 2]
```diff
@@ -124,10 +126,18 @@ private[feature] trait StringIndexerBase extends Params with HasHandleInvalid wi
 
     val outputFields = inputColNames.zip(outputColNames).flatMap {
       case (inputColName, outputColName) =>
-        schema.fieldNames.contains(inputColName) match {
-          case true => Some(validateAndTransformField(schema, inputColName, outputColName))
-          case false if skipNonExistsCol => None
-          case _ => throw new SparkException(s"Input column $inputColName does not exist.")
+        try {
+          val dtype = _transformDataset.col(inputColName).expr.dataType
```
Comment (HyukjinKwon) [inline mllib/src/main/scala/org/apache/spark/ml/feature/StringIndexer.scala:130]: Yeah this LTGM. No need to add `_` in the variable name tho.

Comment (WeichenXu123) [reply mllib/src/main/scala/org/apache/spark/ml/feature/StringIndexer.scala]: renamed .

**Comment by zhengruifeng** [comment]
BTW, is this nested input issue specific to `StringIndexer`? Do other ML implementations also face it?

**Comment by WeichenXu123** [comment]
> BTW, is this nested input issue specific to `StringIndexer`? Do other ML implementations also face it?

Other algorithms might also have it. Will make them supporting this in follow-up PR, but not urgent. Currently user only requires StringIndexer for nested columns.

**Comment by WeichenXu123** [comment]
Merged to master

**Pull request closed without merging**